[BUG] Validation test fail

**Describe the bug**
Validation test fail on commit and PR merge

**To Reproduce**
- Commit or merge a PR
- Verify validation results

**Expected behavior**
Successful validation

**Screenshots**


**Version information**
- Trafiklab integration version: n/a
- Home Assistant version: n/a

**Configuration**
Paste your relevant configuration (redact sensitive data):
```yaml
# Example configuration.yaml
```

**Debug logs**
Paste any relevant log output. Enable debug logging if possible.
```
Run pytest tests/ --cov=custom_components.trafiklab --cov-report=xml
============================= test session starts ==============================
platform linux -- Python 3.11.13, pytest-8.0.2, pluggy-1.6.0
rootdir: /home/runner/work/HomeAssistant_Trafiklab_Integration/HomeAssistant_Trafiklab_Integration
configfile: pyproject.toml
plugins: sugar-1.0.0, syrupy-4.6.1, asyncio-0.23.5, socket-0.7.0, cov-4.1.0, respx-0.20.2, pytest_freezer-0.4.8, unordered-0.5.2, picked-0.5.0, aiohttp-1.0.5, anyio-4.10.0, test-groups-1.0.3, xdist-3.3.1, timeout-2.2.0, homeassistant-custom-component-0.13.109, requests-mock-1.11.0
asyncio: mode=Mode.AUTO
collected 2 items / 1 error

==================================== ERRORS ====================================
__________________ ERROR collecting tests/test_config_flow.py __________________
ImportError while importing test module '/home/runner/work/HomeAssistant_Trafiklab_Integration/HomeAssistant_Trafiklab_Integration/tests/test_config_flow.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.11.13/x64/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_config_flow.py:20: in <module>
    from custom_components.trafiklab.config_flow import (
E   ImportError: cannot import name 'CannotConnect' from 'custom_components.trafiklab.config_flow' (/home/runner/work/HomeAssistant_Trafiklab_Integration/HomeAssistant_Trafiklab_Integration/custom_components/trafiklab/config_flow.py)

---------- coverage: platform linux, python 3.11.13-final-0 ----------
Coverage XML written to file coverage.xml

=========================== short test summary info ============================
ERROR tests/test_config_flow.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.20s ===============================
```

**Additional context**
n/a

--- a/custom_components/trafiklab/config_flow.py
+++ b/custom_components/trafiklab/config_flow.py
@@ -13,6 +13,8 @@ from homeassistant.data_entry_flow import FlowResult
 from homeassistant.exceptions import HomeAssistantError
 import aiohttp
 
+from .api import TrafikLabApiClient, TrafikLabApiError
+
 from .const import (
     DOMAIN,
     CONF_API_KEY,
@@ -80,18 +82,29 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle the initial step."""
-        if user_input is not None:
-            # Store the user input without validation
-            self._api_key = user_input[CONF_API_KEY]
-            self._stop_id = user_input[CONF_STOP_ID]
-            self._name = user_input.get(CONF_NAME, DEFAULT_NAME)
-            
-            # Move to sensor configuration step
-            return await self.async_step_sensor()
+        errors: dict[str, str] = {}
 
-        return self.async_show_form(
-            step_id="user", data_schema=STEP_USER_DATA_SCHEMA
-        )
+        if user_input is not None:
+            try:
+                # Validate before moving on so tests (and users) get feedback.
+                await validate_input(self.hass, user_input)
+            except InvalidApiKey:
+                errors["api_key"] = "Invalid API key"
+            except InvalidStopId:
+                errors["stop_id"] = "Stop not found"
+            except CannotConnect:
+                errors["base"] = "Connection error"
+            except Exception as err:  # pragma: no cover - unexpected
+                _LOGGER.exception("Unexpected exception during validation: %s", err)
+                errors["base"] = "unknown"
+            else:
+                # Persist data for next step
+                self._api_key = user_input[CONF_API_KEY]
+                self._stop_id = user_input[CONF_STOP_ID]
+                self._name = user_input.get(CONF_NAME, DEFAULT_NAME)
+                return await self.async_step_sensor()
+
+        return self.async_show_form(step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors)
 
     async def async_step_sensor(
         self, user_input: dict[str, Any] | None = None
@@ -178,3 +191,47 @@ class OptionsFlowHandler(config_entries.OptionsFlow):
         })
 
         return self.async_show_form(step_id="init", data_schema=schema)
+
+
+# ---------------------------------------------------------------------------
+# Validation helpers & custom exceptions (expected by tests)
+# ---------------------------------------------------------------------------
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidApiKey(HomeAssistantError):
+    """Error to indicate the API key is invalid."""
+
+
+class InvalidStopId(HomeAssistantError):
+    """Error to indicate the stop id is invalid."""
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Returns minimal info dict. Raises custom exceptions on failure.
+    Tests patch this function, so keep signature & return shape stable.
+    """
+    api_key = data[CONF_API_KEY]
+    stop_id = data[CONF_STOP_ID]
+
+    client = TrafikLabApiClient(api_key)
+    try:
+        # Attempt a lightweight departures fetch to validate both key & stop.
+        await client.get_departures(stop_id)
+    except TrafikLabApiError as err:
+        lower = str(err).lower()
+        if "invalid api key" in lower or "authentication failed" in lower:
+            raise InvalidApiKey from err
+        if "not found" in lower or "stop id" in lower:
+            raise InvalidStopId from err
+        raise CannotConnect from err
+    except Exception as err:  # pragma: no cover - network/aio edge case
+        raise CannotConnect from err
+    finally:
+        await client.close()
+
+    # Title used by tests when patched; we mirror behavior.
+    return {"title": data.get(CONF_NAME, DEFAULT_NAME)}

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,0 +1,32 @@
+"""Pytest fixtures for Trafiklab integration tests."""
+from __future__ import annotations
+
+import pytest
+from homeassistant.const import CONF_API_KEY, CONF_STOP_ID, CONF_SENSOR_TYPE, CONF_NAME
+from homeassistant.core import HomeAssistant
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers import entity_registry as er
+
+from custom_components.trafiklab.const import DOMAIN, SENSOR_TYPE_DEPARTURE, DEFAULT_NAME
+
+
+@pytest.fixture
+def mock_config_entry(hass: HomeAssistant) -> ConfigEntry:
+    """Return a mock config entry for the integration."""
+    data = {
+        CONF_API_KEY: "test_api_key",
+        CONF_STOP_ID: "9001",
+        CONF_SENSOR_TYPE: SENSOR_TYPE_DEPARTURE,
+        CONF_NAME: DEFAULT_NAME,
+    }
+    entry = ConfigEntry(
+        version=2,
+        minor_version=0,
+        domain=DOMAIN,
+        title="Trafiklab Test",
+        data=data,
+        source="user",
+        entry_id="test-entry-id",
+    )
+    hass.config_entries._entries.append(entry)  # type: ignore[attr-defined]
+    return entry

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -2,12 +2,17 @@
 from __future__ import annotations
 
 import pytest
-from homeassistant.const import CONF_API_KEY, CONF_STOP_ID, CONF_SENSOR_TYPE, CONF_NAME
 from homeassistant.core import HomeAssistant
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.helpers import entity_registry as er
 
-from custom_components.trafiklab.const import DOMAIN, SENSOR_TYPE_DEPARTURE, DEFAULT_NAME
+from custom_components.trafiklab.const import (
+    DOMAIN,
+    CONF_API_KEY,
+    CONF_STOP_ID,
+    CONF_SENSOR_TYPE,
+    DEFAULT_NAME,
+)
+from custom_components.trafiklab.const import SENSOR_TYPE_DEPARTURE  # separate to keep grouping clear
 
 
 @pytest.fixture
@@ -17,7 +22,7 @@ def mock_config_entry(hass: HomeAssistant) -> ConfigEntry:
         CONF_API_KEY: "test_api_key",
         CONF_STOP_ID: "9001",
         CONF_SENSOR_TYPE: SENSOR_TYPE_DEPARTURE,
-        CONF_NAME: DEFAULT_NAME,
+        "name": DEFAULT_NAME,
     }
     entry = ConfigEntry(
         version=2,
@@ -30,3 +35,9 @@ def mock_config_entry(hass: HomeAssistant) -> ConfigEntry:
     )
     hass.config_entries._entries.append(entry)  # type: ignore[attr-defined]
     return entry
+
+
+@pytest.fixture(autouse=True)
+def _auto_enable_custom_integrations(enable_custom_integrations):  # noqa: D401
+    """Automatically enable loading of custom_components directory."""
+    yield

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4,6 +4,7 @@ from __future__ import annotations
 import pytest
 from homeassistant.core import HomeAssistant
 from homeassistant.config_entries import ConfigEntry
+from pytest_homeassistant_custom_component.common import MockConfigEntry
 
 from custom_components.trafiklab.const import (
     DOMAIN,
@@ -16,24 +17,22 @@ from custom_components.trafiklab.const import SENSOR_TYPE_DEPARTURE  # separate 
 
 
 @pytest.fixture
-def mock_config_entry(hass: HomeAssistant) -> ConfigEntry:
-    """Return a mock config entry for the integration."""
+def mock_config_entry(hass: HomeAssistant) -> ConfigEntry:  # type: ignore[override]
+    """Provide a mock config entry added to hass."""
     data = {
         CONF_API_KEY: "test_api_key",
         CONF_STOP_ID: "9001",
         CONF_SENSOR_TYPE: SENSOR_TYPE_DEPARTURE,
         "name": DEFAULT_NAME,
     }
-    entry = ConfigEntry(
-        version=2,
-        minor_version=0,
+    entry = MockConfigEntry(
         domain=DOMAIN,
-        title="Trafiklab Test",
         data=data,
-        source="user",
-        entry_id="test-entry-id",
+        title="Trafiklab Test",
+        version=2,
+        unique_id="9001_departure",
     )
-    hass.config_entries._entries.append(entry)  # type: ignore[attr-defined]
+    entry.add_to_hass(hass)
     return entry
 
 

